src/mono/mono/mini/mini-s390x.c

```diff
@@ -4834,606 +5322,461 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
+            break;
+        case OP_VFLCDB:
+            s390_vflcdb (code, ins->dreg, ins->sreg1);
+            break;
+        case OP_VFLCSB:
+            s390_vflcsb (code, ins->dreg, ins->sreg1);
+            break;
+        case OP_INSERT_I1:
+            s390_vlr (code, ins->dreg, ins->sreg1);
+            s390_vlvgb (code, ins->dreg, ins->sreg2, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_INSERT_I2:
+            s390_vlr (code, ins->dreg, ins->sreg1);
+            s390_vlvgh (code, ins->dreg, ins->sreg2, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_INSERT_I4:
+            s390_vlr (code, ins->dreg, ins->sreg1);
+            s390_vlvgf (code, ins->dreg, ins->sreg2, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_INSERT_I8:
+            s390_vlr (code, ins->dreg, ins->sreg1);
+            s390_vlvgg (code, ins->dreg, ins->sreg2, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_INSERT_R4:
+            s390_vlgvf (code, s390_r13, ins->sreg2, 0, 0);
+            s390_vlvgf (code, ins->dreg, s390_r13, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_INSERT_R8:
+            s390_vlgvg (code, s390_r13, ins->sreg2, 0, 0);
+            s390_vlvgg (code, ins->dreg, s390_r13, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_EXTRACT_I1:
+            s390_vlgvb (code, ins->dreg, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_EXTRACT_I2:
+            s390_vlgvh (code, ins->dreg, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_EXTRACT_I4:
+            s390_vlgvf (code, ins->dreg, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_EXTRACT_I8:
+            s390_vlgvg (code, ins->dreg, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            break;
+        case OP_EXTRACT_R4:
+            s390_vlgvf (code, s390_r13, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            s390_vlvgf (code, ins->dreg, s390_r13, 0, 0);
+            break;
+        case OP_EXTRACT_R8:
+            s390_vlgvg (code, s390_r13, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+            s390_vlvgg (code, ins->dreg, s390_r13, 0, 0);
+            break;
+        case OP_XEXTRACT_I1:
+            s390_vlgvb (code, ins->dreg, ins->sreg1, ins->sreg2, 0);
+            break;
+        case OP_XEXTRACT_I2:
+            s390_vlgvh (code, ins->dreg, ins->sreg1, ins->sreg2, 0);
+            break;
+        case OP_XEXTRACT_I4:
+            s390_vlgvf (code, ins->dreg, ins->sreg1, ins->sreg2, 0);
+            break;
+        case OP_XEXTRACT_I8:
+            s390_vlgvg (code, ins->dreg, ins->sreg1, ins->sreg2, 0);
+            break;
+        case OP_XEXTRACT_R4:
+            s390_vlgvf (code, s390_r13, ins->sreg1, ins->sreg2, 0);
+            s390_ldgr (code, ins->dreg, s390_r13);
```

Comment:
This puts it into the wrong half of the FPR, doesn't it?  Should probably just use the same code as `OP_XEXTRACT_R8` ...